src/porepy/grids/grid_bucket.py

```diff
@@ -728,6 +728,85 @@ def eliminate_node(self, node):
 
 #------------------------------------------------------------------------------#
 
+    def loop_nodes(self, fct):
+        """
+        Loop on all the nodes and evaluate a function on each of them.
+
+        Parameter:
+            fct: function to evaluate. It takes a grid and the related data and
+                returns a scalar.
+
+        Returns:
+            values: vector containing the function evaluated on each node,
+                ordered by 'node_number'.
+
+        """
+        values = np.empty(self.size())
+        for g, d in self:
+            values[d['node_number']] = fct(g, d)
+        return values
+
+#------------------------------------------------------------------------------#
+
+    def loop_edges(self, fct):
+        """
+        Loop on all the edges and evaluate a function on each of them.
+
+        Parameter:
+            fct: function to evaluate. It returns a scalar and takes: the higher
+                and lower dimensional grids, the higher and lower dimensional
+                data, the global data.
+
+        Returns:
+            matrix: sparse strict upper triangular matrix containing the function
+                evaluated on each edge (pair of nodes), ordered by their
+                relative 'node_number'.
+
+        """
+        i = np.zeros(self.graph.number_of_edges(), dtype=int)
+        j = np.zeros(i.size, dtype=int)
+        values = np.zeros(i.size)
+
+        # Loop over the edges of the graph (pair of connected nodes)
+        idx = 0
+        for e, data in self.edges_props():
+            g_l, g_h = self.sorted_nodes_of_edge(e)
+            data_l, data_h = self.node_props(g_l), self.node_props(g_h)
```

Comment:
Well, then I won't blame you. Nevertheless, it is hard to read code when you have to count the number of commas on each side of the equality sign. I much prefer an extra line or two.